File: benchmarks/multilang-wasm/dom-table-sort-filter-pagination/table_sort_kernel.ts
// table_sort_kernel.ts — AssemblyScript multilang compute core for
// dom.table-sort-filter-pagination.v1. Same ABI as the C kernel: generates
// the frozen 120-action trace from seed 0x31415926, runs the 5,000-row JS
// reference model, writes counters to a fixed offset, returns pageScoreSum.
// Raw linear-memory access only (no heap allocation, no runtime imports).

const ROWS = 5000;
const PAGE = 50;
const ACTIONS = 120;
const SCORES_OFFSET: usize = 0; // i32[5000] → bytes 0..20000
const FILT_OFFSET: usize = 20000; // i32[5000] → bytes 20000..40000
const RESULTS_OFFSET: usize = 40000; // u32[5] → bytes 40000..40020
// Two 8-byte scratch buffers for name-column string comparison (no heap alloc).
const NAME_A_OFFSET: usize = 40032;
const NAME_B_OFFSET: usize = 40048;

let seed: u32 = 0;

function randNext(): f64 {
  seed ^= seed << 13;
  // replicate the JS engine's rand(): >> 17 applies to the int32
  // interpretation (arithmetic, sign-extending).
  seed ^= (<i32> seed >> 17) as u32;
  seed ^= seed << 5;
  return (<f64> seed) / 4294967296.0;
}

function catRank(c: i32): i32 {
  if (c === 0) return 0;
  if (c === 1) return 1;
  if (c === 2) return 4;
  if (c === 3) return 2;
  return 3;
}
function statRank(s: i32): i32 {
  if (s === 0) return 0;
  if (s === 1) return 2;
  return 1;
}
function cmpName(a: i32, b: i32): i32 {
  let na: i32 = 0;
  let nb: i32 = 0;
  let x = a;
  do {
    store<u8>(NAME_A_OFFSET + <usize> na, <u8> (48 + (x % 10)));
    na++;
    x = (x / 10) as i32;
  } while (x !== 0);
  x = b;
  do {
    store<u8>(NAME_B_OFFSET + <usize> nb, <u8> (48 + (x % 10)));
    nb++;
    x = (x / 10) as i32;
  } while (x !== 0);
  let i: i32 = 0;
  while (i < na && i < nb) {
    const ca = load<u8>(NAME_A_OFFSET + <usize> (na - 1 - i));
    const cb = load<u8>(NAME_B_OFFSET + <usize> (nb - 1 - i));
    if (ca !== cb) return ca < cb ? -1 : 1;
    i++;
  }
  if (na !== nb) return na < nb ? -1 : 1;
  return 0;
}
function cmpRow(aId: i32, bId: i32, col: i32, asc: i32): i32 {
  let cmp: i32 = 0;
  if (col === 0) {
    cmp = aId < bId ? -1 : (aId > bId ? 1 : 0);
  } else if (col === 1) {
    cmp = cmpName(aId, bId);
  } else if (col === 2) {
    const ra = catRank(aId % 5);
    const rb = catRank(bId % 5);
    cmp = ra < rb ? -1 : (ra > rb ? 1 : 0);
  } else if (col === 3) {
    const sa = load<i32>(SCORES_OFFSET + <usize> aId * 4);
    const sb = load<i32>(SCORES_OFFSET + <usize> bId * 4);
    cmp = sa < sb ? -1 : (sa > sb ? 1 : 0);
  } else if (col === 4) {
    const ra = statRank(aId % 3);
    const rb = statRank(bId % 3);
    cmp = ra < rb ? -1 : (ra > rb ? 1 : 0);
  }
  return asc !== 0 ? cmp : -cmp;
}

export function table_sort_trace(): i32 {
  for (let i = 0; i < ROWS; i++) {
    store<i32>(SCORES_OFFSET + i * 4, (i * 37) % 1000);
    store<i32>(FILT_OFFSET + i * 4, i);
  }
  let filteredCount: i32 = ROWS;
  let currentPage: i32 = 0;
  let pageSize: i32 = PAGE;
  let totalSorts: u32 = 0;
  let totalFilters: u32 = 0;

  seed = 0x31415926;
  for (let a = 0; a < ACTIONS; a++) {
    const opType = randNext();
    if (opType < 0.35) {
      const col = <i32> (randNext() * 5.0);
      const asc: i32 = randNext() > 0.5 ? 1 : 0;
      for (let i = 1; i < filteredCount; i++) {
        const key = load<i32>(FILT_OFFSET + i * 4);
        let j = i - 1;
        while (j >= 0 && cmpRow(key, load<i32>(FILT_OFFSET + j * 4), col, asc) < 0) {
          store<i32>(FILT_OFFSET + (j + 1) * 4, load<i32>(FILT_OFFSET + j * 4));
          j--;
        }
        store<i32>(FILT_OFFSET + (j + 1) * 4, key);
      }
      totalSorts += 1;
    } else if (opType < 0.70) {
      const fIdx = <i32> (randNext() * 6.0);
      let out: i32 = 0;
      if (fIdx === 5) {
        for (let i = 0; i < ROWS; i++) {
          store<i32>(FILT_OFFSET + out * 4, i);
          out++;
        }
      } else {
        const targetCat = fIdx;
        for (let i = 0; i < ROWS; i++) {
          if ((i % 5) === targetCat) {
            store<i32>(FILT_OFFSET + out * 4, i);
            out++;
          }
        }
      }
      filteredCount = out;
      totalFilters += 1;
    } else if (opType < 0.90) {
      const page = <i32> (randNext() * 20.0);
      currentPage = page;
      pageSize = PAGE;
    } else {
      const rowId = <i32> (randNext() * 5000.0);
      const newScore = <i32> (randNext() * 1000.0);
      if (rowId >= 0 && rowId < ROWS) {
        store<i32>(SCORES_OFFSET + rowId * 4, newScore);
      }
    }
  }

  const pageStart = currentPage * pageSize;
  let pageEnd = pageStart + pageSize;
  if (pageEnd > filteredCount) pageEnd = filteredCount;
  let sliceLen = pageEnd - pageStart;
  if (sliceLen < 0) sliceLen = 0;
  let pageScoreSum: u32 = 0;
  for (let i = pageStart; i < pageEnd; i++) {
    const id = load<i32>(FILT_OFFSET + i * 4);
    pageScoreSum += <u32> load<i32>(SCORES_OFFSET + id * 4);
  }

  store<u32>(RESULTS_OFFSET, <u32> filteredCount);
  store<u32>(RESULTS_OFFSET + 4, totalSorts);
  store<u32>(RESULTS_OFFSET + 8, totalFilters);
  store<u32>(RESULTS_OFFSET + 12, <u32> sliceLen);
  store<u32>(RESULTS_OFFSET + 16, pageScoreSum);
  return <i32> pageScoreSum;
}

File: benchmarks/multilang-wasm/dom-table-sort-filter-pagination/table_sort_kernel.c
// table_sort_kernel.c — multilang compute core for
// dom.table-sort-filter-pagination.v1.
//
// ABI (mirrors benchmarks/multilang-wasm/dom-grid-movement/grid_kernel.c): the
// kernel GENERATES the frozen 120-action trace internally from the pinned
// seed (0x31415926), runs the JS reference model (runTableSortFilterJS: 5,000
// rows with category=i%5, status=i%3, score=(i*37)%1000; stable sort with JS
// string-order ranks for category/status/name; filter resets to id-order;
// edit_cell mutates score at rowId; paginate updates current page/size),
// writes counters to a FIXED memory offset, returns pageScoreSum.
//
// Results (fixed offset 40000):
//   [0] filteredCount [1] totalSorts [2] totalFilters
//   [3] pageSliceCount [4] pageScoreSum
// Exports: i32 table_sort_trace() -> pageScoreSum

#define ROWS 5000
#define PAGE 50
#define ACTIONS 120
#define RES_OFFSET 40000     // u32[5] (past the AS-shared 40000-byte data region)

static unsigned int seed = 0;
static double rand_next(void) {
  seed ^= seed << 13;
  // the engine's JS rand() applies >> 17 to the int32 interpretation
  // (arithmetic, sign-extending) — replicate exactly.
  seed ^= (unsigned int)((int)seed >> 17);
  seed ^= seed << 5;
  return (double)seed / 4294967296.0;
}

// cols: 0 id, 1 name, 2 category, 3 score, 4 status
// JS Array.sort compares as strings for category / status / name (alphabetical
// on the underlying string labels). Fold to arithmetic ranks so we can sort
// integer keys with stable insertion sort.
// categories: alpha(0) < beta(1) < delta(3) < epsilon(4) < gamma(2)
static int cat_rank(int c) {
  if (c == 0) return 0; // alpha
  if (c == 1) return 1; // beta
  if (c == 2) return 4; // gamma
  if (c == 3) return 2; // delta
  return 3;             // epsilon
}
// statuses: active(0) < archived(2) < pending(1)
static int stat_rank(int s) {
  if (s == 0) return 0; // active
  if (s == 1) return 2; // pending
  return 1;             // archived
}

// JS compares names as strings: name = "User " + id. Prefix is common; the
// comparison reduces to lexicographic comparison of the decimal digit string
// of the id (e.g. "User 10" < "User 9" because '1' < '9').
static int cmp_name(int a, int b) {
  char sa[8], sb[8];
  int na = 0, nb = 0;
  int x = a;
  do { sa[na++] = (char)('0' + (x % 10)); x /= 10; } while (x);
  x = b;
  do { sb[nb++] = (char)('0' + (x % 10)); x /= 10; } while (x);
  int i = 0;
  while (i < na && i < nb) {
    const char ca = sa[na - 1 - i];
    const char cb = sb[nb - 1 - i];
    if (ca != cb) return ca < cb ? -1 : 1;
    i++;
  }
  if (na != nb) return na < nb ? -1 : 1;
  return 0;
}

// Returns -1 / 0 / +1 for row a vs row b under the JS comparator (col+asc).
static int cmp_row(const int *scores, int a_id, int b_id, int col, int asc) {
  int cmp = 0;
  if (col == 0) { // id
    cmp = a_id < b_id ? -1 : (a_id > b_id ? 1 : 0);
  } else if (col == 1) { // name
    cmp = cmp_name(a_id, b_id);
  } else if (col == 2) { // category
    const int ra = cat_rank(a_id % 5);
    const int rb = cat_rank(b_id % 5);
    cmp = ra < rb ? -1 : (ra > rb ? 1 : 0);
  } else if (col == 3) { // score
    const int sa = scores[a_id];
    const int sb = scores[b_id];
    cmp = sa < sb ? -1 : (sa > sb ? 1 : 0);
  } else if (col == 4) { // status
    const int ra = stat_rank(a_id % 3);
    const int rb = stat_rank(b_id % 3);
    cmp = ra < rb ? -1 : (ra > rb ? 1 : 0);
  }
  return asc ? cmp : -cmp;
}

__attribute__((export_name("table_sort_trace")))
int table_sort_trace(void) {
  int scores[ROWS];
  int filt[ROWS];
  unsigned int *results = (unsigned int *)RES_OFFSET;

  for (int i = 0; i < ROWS; i++) {
    scores[i] = (i * 37) % 1000;
    filt[i] = i;
  }
  int filtered_count = ROWS;
  int current_page = 0;
  int page_size = PAGE;
  unsigned int total_sorts = 0;
  unsigned int total_filters = 0;

  seed = 0x31415926;
  for (int a = 0; a < ACTIONS; a++) {
    const double op_type = rand_next();
    if (op_type < 0.35) { // sort
      // JS generator: col index = floor(rand()*5), asc = rand() > 0.5
      const int col = (int)(rand_next() * 5.0);
      const int asc = rand_next() > 0.5 ? 1 : 0;
      // stable insertion sort over filt[0..filtered_count) using cmp_row
      for (int i = 1; i < filtered_count; i++) {
        const int key = filt[i];
        int j = i - 1;
        while (j >= 0 && cmp_row(scores, key, filt[j], col, asc) < 0) {
          filt[j + 1] = filt[j];
          j--;
        }
        filt[j + 1] = key;
      }
      total_sorts++;
    } else if (op_type < 0.70) { // filter
      // JS generator: filter index = floor(rand()*6); filters =
      //   ["alpha","beta","gamma","delta","epsilon",""]. Empty query resets to
      //   all rows (in id order); named query keeps rows whose category matches.
      const int f_idx = (int)(rand_next() * 6.0);
      int out = 0;
      if (f_idx == 5) {
        for (int i = 0; i < ROWS; i++) filt[out++] = i;
      } else {
        // filters[f_idx] maps to category code by name — but the JS filter is
        // `r.category.includes(q) || r.name.toLowerCase().includes(q)`. For
        // the 5 real categories the substring `q === r.category`, and the
        // name "user N" never contains any of these queries. So filter by
        // exact category-code equality:
        //   filters[0]="alpha"→cat 0, [1]="beta"→1, [2]="gamma"→2,
        //   [3]="delta"→3, [4]="epsilon"→4.
        const int target_cat = f_idx;
        for (int i = 0; i < ROWS; i++) {
          if ((i % 5) == target_cat) filt[out++] = i;
        }
      }
      filtered_count = out;
      total_filters++;
    } else if (op_type < 0.90) { // paginate
      // JS generator: page = floor(rand()*20), pageSize = 50
      const int page = (int)(rand_next() * 20.0);
      current_page = page;
      page_size = PAGE;
      (void)current_page;
    } else { // edit_cell
      const int row_id = (int)(rand_next() * 5000.0);
      const int new_score = (int)(rand_next() * 1000.0);
      if (row_id >= 0 && row_id < ROWS) scores[row_id] = new_score;
    }
  }

  // Compute final page slice
  const int page_start = current_page * page_size;
  int page_end = page_start + page_size;
  if (page_end > filtered_count) page_end = filtered_count;
  int slice_len = page_end - page_start;
  if (slice_len < 0) slice_len = 0;
  unsigned int page_score_sum = 0;
  for (int i = page_start; i < page_end; i++) {
    page_score_sum += (unsigned int)scores[filt[i]];
  }

  results[0] = (unsigned int)filtered_count;
  results[1] = total_sorts;
  results[2] = total_filters;
  results[3] = (unsigned int)slice_len;
  results[4] = page_score_sum;
  return (int)page_score_sum;
}
